File: ENT3C.jl
function parse_args(args)
    installs = nothing
    resolve_env = nothing
    config_file = nothing
    julia_version = nothing
    for arg in args
        if startswith(arg, "--config-file=")
            config_file = split(arg, "=")[2]
        end
        if startswith(arg, "--install-deps=")
            installs = split(arg, "=")[2]
        end
        if startswith(arg, "--resolve-env=")
            resolve_env = split(arg, "=")[2]
        end
        if startswith(arg, "--julia-version=")
            julia_version = split(arg, "=")[2]
        end
    end

    if isnothing(config_file) || config_file == ""
        throw(ArgumentError("--config-file is missing or empty"))
    end

    return installs, resolve_env, config_file, julia_version
end

function check_missing_packages(required_packages)
        installed_packages = [info.name for info in values(Pkg.dependencies())]
        missing_packages = [pkg for pkg in required_packages if !(pkg in installed_packages)]
        return missing_packages
end


function install_packages(packages)
        for pkg in packages
                try
                        Pkg.add(pkg)
                catch e
                        println("Failed to install package $pkg: $e")
                        exit(1)
                end
        end
end

using Pkg
installs, resolve_env, config_file, julia_version = parse_args(ARGS)
#args = ["--config-file=config/config.json", "--install-deps=no"]
#installs, config_file = parse_args(args)

required_packages = ["DataFrames", "BenchmarkTools", "JSON", "Printf", "Plots", "ColorSchemes", "SuiteSparse", "HDF5", "NaNStatistics", "Statistics", "Combinatorics", "CSV"]
missing_packages = check_missing_packages(required_packages)

if isnothing(installs) && isnothing(resolve_env) && !isempty(missing_packages)
    println("Error! Missing julia packages. Please specify one of the following:")
    println("1. Use --install-deps=yes to install the packages globally.")
    println("   Or")
    println("2. Specify --resolve-env=yes with --julia-version=1.10.4 or 1.11.2 to install and resolve environment.")
    exit(1)
end

if installs == "yes" && length(missing_packages)!=1 && resolve_env != "yes"
    println("Installing missing dependencies globally.")
    println(resolve_env)
    install_packages(missing_packages)
end


if resolve_env == "yes"
        if isnothing(julia_version) || julia_version == ""
           throw(ArgumentError("--resolve-env requires julia version specification --julia-version=1.11.2 or --julia-version=1.10.4"))
        end
        println("Resolving ENT3C julia enviornment.")
        Pkg.activate("project_files/$(julia_version)")
        Pkg.add(["DataFrames", "BenchmarkTools", "JSON", "Printf", "Plots", "ColorSchemes", "SuiteSparse", "HDF5", "NaNStatistics", "Statistics", "Combinatorics", "CSV"])
        Pkg.instantiate()
end


using DataFrames, BenchmarkTools, JSON, Printf, Plots, ColorSchemes, SuiteSparse

include("JULIA_functions/ent3c_functions.jl")
######################################################
# user input config file
######################################################
#config = JSON.parsefile("config/config.test.json")

function get_config(config_file)
    config=nothing
    if config_file === nothing
        println("Please specify a configuration file.")
    else
    println("Using config file: $config_file")
    try
        config = JSON.parsefile(config_file)
        println("Config data: ", config)
    catch e
        println("Error reading config file: ", e)
    end
    end
    return config
end
config = get_config(config_file)
#########################################################

SUB_M_SIZE_FIX = config["SUB_M_SIZE_FIX"]
PHI_MAX = config["PHI_MAX"]
Resolutions = config["Resolution"]
Resolutions = split(Resolutions, ",")
Resolutions = [parse(Float64, res) |> Int for res in Resolutions]
phi = config["phi"]
NormM = config["NormM"]
DATA_PATH = config["DATA_PATH"]
FILES = config["FILES"]
Biological_replicates = any(x -> contains(x, '_'), FILES[2, :])
FILES = hcat(joinpath.(DATA_PATH, FILES[1:2:end-1]), FILES[2:2:end])
OUT_PREFIX = config["OUT_PREFIX"]
CHRSPLIT = config["CHRSPLIT"]
weights_name = config["WEIGHTS_NAME"]

if OUT_PREFIX === nothing
    OUT_PREFIX = @sprintf("%dkb",Resolution/1e3)
end
OUT_DIR = config["OUT_DIR"]
OUT_DIR = string(OUT_DIR,"JULIA")
if !isdir(OUT_DIR)
    mkpath(OUT_DIR)
end

ChrNrs = string(config["ChrNr"])
if contains(ChrNrs, ",")
    ChrNrs = split(ChrNrs, ",")
else
    ChrNrs = [ChrNrs]
end
#############################################################################
# ENT3C table
##############################################################################
ENT3C_OUT = main(FILES,Resolutions,ChrNrs,SUB_M_SIZE_FIX,CHRSPLIT,PHI_MAX,phi,NormM,weights_name)
#f = [filter(row -> isnan(row.S)  , ENT3C_OUT).binNrStart[1], filter(row -> isnan(row.S)  , ENT3C_OUT).binNrEnd[1]]
#filter(row -> row.binNrStart==f[1] && row.binNrEnd==f[2], ENT3C_OUT)
#############################################################################
# similarity table
#############################################################################
SAMPLES::Array=unique(ENT3C_OUT.Name)
if length(SAMPLES)>1 
        Similarity = get_similarity_table(ENT3C_OUT,ChrNrs,Resolutions,Biological_replicates)
        CSV.write(@sprintf("%s/%s_ENT3C_similarity.csv",OUT_DIR,OUT_PREFIX), Similarity)
end
CSV.write(@sprintf("%s/%s_ENT3C_OUT.csv",OUT_DIR,OUT_PREFIX), ENT3C_OUT)
	

File: ENT3C.m
clear all;close all
addpath('MATLAB_functions/')
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% load json file
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
config = fileread('config/config.json');
config = jsondecode(config);
PHI_MAX=config.PHI_MAX;
CHRSPLIT=config.CHRSPLIT;
SUB_M_SIZE_FIX=config.SUB_M_SIZE_FIX;
Resolutions=str2num(config.Resolution);
phi=config.phi;
NormM=config.NormM;
DATA_PATH=config.DATA_PATH;
FILES = config.FILES;
FILES = reshape(FILES,2,size(FILES,1)/2);
FILES = [cellfun(@(file) fullfile(DATA_PATH, file), FILES(1,:), 'UniformOutput', false); FILES(2,:)];
if any(contains(FILES(2,:),'_'))% only important in case of computing mean(Q_BRs) mean(Q_nonBRs)
    Biologicap_replicates=true;
else
    Biologicap_replicates=false;
end
OUT_DIR=config.OUT_DIR;
OUT_DIR=[OUT_DIR,'MATLAB'];
OUT_PREFIX=config.OUT_PREFIX;
if isempty(OUT_PREFIX)
    OUT_PREFIX=sprintf('%dkb',Resolution/1e3);
end
if ~exist(OUT_DIR, 'dir')
    mkdir(OUT_DIR);
end
ChrNrs=str2num(config.ChrNr);

ENT3C_OUT=[];
for Resolution=Resolutions
    for ChrNr=ChrNrs

        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % load files
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        INFO = struct('FN', '', 'META', '');
        FNs = [];
        for f = 1:size(FILES,2)
            FNs = [FNs, INFO];
            FNs(end).FN = FILES{1,f};
            FNs(end).META = FILES{2,f};
        end
        CM=jet(size(FNs,2));
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % extract common empty bins of input matrices
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        EXCLUDE=[];
        for f = 1:numel(FNs)
            FN=FNs(f).FN;
            [M,BIN_TABLE]=load_cooler(FN,ChrNr,Resolution,NormM);
            if NormM==0
                EXCLUDE=[EXCLUDE;BIN_TABLE.binNr(isnan(BIN_TABLE.CONTACT))];
                    elseif NormM==1
                EXCLUDE=[EXCLUDE;BIN_TABLE.binNr(isnan(BIN_TABLE.CONTACT));BIN_TABLE.binNr(isnan(BIN_TABLE.weights))];
            end
        end
        EXCLUDE=unique(EXCLUDE);
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % fill ENT3C table
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        for f = 1:numel(FNs)
            FN=FNs(f).FN;

            [M,BIN_TABLE]=load_cooler(FN,ChrNr,Resolution,NormM);
            INCLUDE = 1:size(M,1);
            INCLUDE = setdiff(INCLUDE,EXCLUDE);

            M = M(INCLUDE,INCLUDE);
            BIN_TABLE = BIN_TABLE(INCLUDE,:);

            writetable(BIN_TABLE,sprintf('%s_chr%d_BINMATRIXMATLAB.csv',FN,ChrNr),'Delimiter','tab')

            [S, SUB_M_SIZE1, WN1, WS1, BIN_TABLE_NEW] = vN_entropy(M,SUB_M_SIZE_FIX,CHRSPLIT,PHI_MAX,phi,BIN_TABLE);
            N = length(S);

            OUT1 = table(repmat({FNs(f).META},N,1),...
                repmat(ChrNr,N,1),repmat(Resolution,N,1),...
                repmat(SUB_M_SIZE1,N,1),repmat(WN1,N,1),repmat(WS1,N,1),...
                BIN_TABLE_NEW(:,1),BIN_TABLE_NEW(:,2),...
                BIN_TABLE_NEW(:,3),BIN_TABLE_NEW(:,4),S,...
                'VariableNames',{'Name','ChrNr','Resolution','sub_m_dim','WN','phi',...
                'binNrStart','binNrEND','START','END','S'});

            ENT3C_OUT=[ENT3C_OUT;OUT1];
            ENT3C_OUT
        end
    end
end

writetable(ENT3C_OUT,sprintf('%s/%s_ENT3C_OUT.csv',OUT_DIR,OUT_PREFIX),'Delimiter','tab')

%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% fill similarity table
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

SAMPLES=unique(ENT3C_OUT.Name);

Similarity = [];
if numel(SAMPLES)>1
    for Resolution=Resolutions
        figure(Resolution);hold on;tiledlayout('flow');hold on
        for ChrNr=ChrNrs
            figure(Resolution);nexttile

            comparisons = get_pairwise_combs(SAMPLES);

            plotted = "tempstring";
            c=1;
            for f=1:size(comparisons,1)

                S1 = ENT3C_OUT(strcmp(ENT3C_OUT.Name,comparisons{f,1})&ENT3C_OUT.ChrNr==ChrNr&ENT3C_OUT.Resolution==Resolution,:);
                S2 = ENT3C_OUT(strcmp(ENT3C_OUT.Name,comparisons{f,2})&ENT3C_OUT.ChrNr==ChrNr&ENT3C_OUT.Resolution==Resolution,:);
                Q = corrcoef(S1.S,S2.S);Q=Q(1,2);

                Similarity=[Similarity;...
                    table(Resolution,ChrNr,comparisons{f,1},comparisons{f,2},Q,...
                    'VariableNames',{'Resolution','ChrNr','Sample1','Sample2','Q'})];

                if ~ismember(comparisons{f,1}, plotted)
                    figure(Resolution);plot(S1.S,'color',CM(c,:),'linewidth',1.3);hold on;axis tight
                    plotted = [plotted;comparisons{f,1}];c=c+1;
                end
                if ~ismember(comparisons{f,2}, plotted)
                    figure(Resolution);plot(S2.S,'color',CM(c,:),'linewidth',1.3);hold on;axis tight
                    plotted = [plotted;comparisons{f,2}];c=c+1;
                end
            end
            if Biologicap_replicates
                title(sprintf('Chr%d %dkb\n$\\overline{Q}_{BR}=%4.2f$ $\\overline{Q}_{nonBR}=%4.2f$',...
                    ChrNr,Resolution/1e3,...
                    mean(Similarity.Q(Similarity.ChrNr==ChrNr&Similarity.Resolution==Resolution&...
                    strcmp(extractBefore(Similarity.Sample1,'_'),extractBefore(Similarity.Sample2,'_')))),...
                    mean(Similarity.Q(Similarity.ChrNr==ChrNr&Similarity.Resolution==Resolution&...
                    ~strcmp(extractBefore(Similarity.Sample1,'_'),extractBefore(Similarity.Sample2,'_'))))),...
                    'interpreter','latex','fontsize',10)
            else
                title(sprintf('Chr%d %dkb',ChrNr,Resolution/1e3),'interpreter','latex','fontsize',9)
            end

            if ChrNr==ChrNrs(end)
                plotted(1)=[];plotted=strrep(plotted,'_',' ');
                legend(plotted,'location','northeastoutside')
            end
            set(gca,'FontSize',9)
            plotted
        end

        figure(Resolution);
        set(gcf,'Position',[52 215 1717 813])
        saveas(gcf,sprintf('%s/%s_ENT3C_signals.svg',OUT_DIR,OUT_PREFIX))

    end
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    % save output
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    writetable(Similarity,sprintf('%s/%s_ENT3C_similarity.csv',OUT_DIR,OUT_PREFIX),'Delimiter','tab')
end
